rq/cli/helpers.py

```diff
@@ -198,6 +200,74 @@ def setup_loghandlers_from_args(verbose, quiet, date_format, log_format):
     setup_loghandlers(level, date_format=date_format, log_format=log_format)
 
 
+def job_func(func, args, kwargs):
+    try:
+        importlib.import_module(func)  # execute file
+    except ImportError:
+        return import_attribute(func)(*args, **kwargs)
+
+
+def cast_element(element, arg_type):
+    if arg_type == bool:
```

Comment:
When do we need to cast booleans from CLI arguments? I think in CLI context we should use something like `--at-front` so we shouldn't need to parse `true` or `false`

Reply:
`cast_element` is not used for something like `--at-front`, but is used [here](https://github.com/rq/rq/blob/809395edbc10c3d46f7477937abe6e80a388479e/rq/cli/helpers.py#L244) for casting the args of the function you want to call.

```
rq enqueue path.to.func --boolean true
```

will call `path.to.func` with the boolean `True` as arg.

Reply:
So what I mean is that we shouldn't have a CLI argument with this kind of signature.

* `rq enqueue path.to.func --at-front true`
* `rq enqueue path.to.func --at-front false`

The two commands above can be better expressed as `rq enqueue path.to.func --at-front`

So I don't want to support a use case where the user passed a boolean as CLI argument.

Reply:
I have implemented `--at-front` and `--quiet` as simple click.option with `is_flag=true`.

So if you want to enqueue at front you can call:

```
eq enqueue path.to.func --at-front
```

If you want to enqueue at front with `False` as an argument, you can call:

```
eq enqueue path.to.func --at-front --bool false
```

Reply:
You shouldn't need to call `rq enqueue path.to.func --at-front --bool false` because this is the default behavior.

So if you want to enqueue at front you can call `enqueue path.to.func --at-front` (like you mentioned). Otherwise it should be simply `rq enqueue path.to.func`, no boolean argument necessary.

Reply:
`--boolean` is used for the same as `--int` or `--float`.

`--boolean` is used if you want to add an argument to the function you want to enqueue.

How you can see here `--at-front` is a click option and `--boolean` has no effect to it.

https://github.com/rq/rq/blob/2741c3bf50d4b2a2fae5051690a2c31d552c68b3/rq/cli/cli.py#L322

Here a simple example:

Python:
```python
def func(arg=None):  # path.to.func
    print(arg)
    print(type(arg))
```


Then execute:

```
$ rq enqueue path.to.func --boolean false
```
And it returns:
```
12:14:18 default: path.to.func(False) (97fbe646-4414-427d-9dac-0e3b136d9208)
False
<class 'bool'>
12:14:18 default: Job OK (97fbe646-4414-427d-9dac-0e3b136d9208)
12:14:18 Result is kept for 500 second
```

So you see, that the argument is `False` as a boolean.

But executing:
```
$ rq enqueue path.to.func
```
returns:
```
12:15:58 default: path.to.func() (6be872b6-aaf6-4cf7-b354-84c7229b9827)
None
<class 'NoneType'>
12:15:58 default: Job OK (6be872b6-aaf6-4cf7-b354-84c7229b9827)
12:15:58 Result is kept for 500 seconds
```

So this time there is no argument so the python functions is called without any args and `None`, specifiered in the header of the function, is taken as the default arg of the function.

For all this examples you could add `--at-front` without anything after it to enqueue the job at front.

Reply:
Sorry for being dense, now I get it. So the problem is actually passing args/kwargs to the callable itself.

Are you ok if we merge this in without the capability to pass in args/kwargs first to the callable?

If we do support this, I want our CLI API to be similar to https://httpie.io . Take a look at how they handle [non string JSON fields](https://httpie.io/docs#non-string-json-fields).

I'm not sure how easy or difficult it would be to achieve though.

Reply:
Ok, I implement:

<table><thead><tr><th></th><th>keyword</th><th>no keyword</th></tr></thead><tbody><tr><td>string</td><td rowspan="2">Like <a href="https://httpie.io/docs#non-string-json-fields">this</a></td><td>Just the value</td></tr><tr><td>json</td><td><code>:value</code><br>(e.g. <code>:1</code> would be <code>1</code> as an integer.)</td></tr></tbody></table>

Reply:
With `\` as escape char

Reply:
So `:` and `=` are not allowed to be in keywords

Reply:
One way we can go about this is to use [ast.literal_eval()](https://docs.python.org/3/library/ast.html#ast.literal_eval) to safely evalute inputs.

`literal_eval()` supports strings, bytes, numbers, tuples, lists, dicts, sets, booleans, and None so this should be good enough. So we basically support whatever `literal_eval()` supports.

Reply:
Do you mind documenting this feature too so people know how to use it? The docs are under `/docs` directory.

Reply:
Thinking further about this, can we just pass every args/kwargs passed to `enqueue` via `literal_eval()`? So we can avoid the unfamiliar `:=` altogether.

Reply:
The goal is to be able to do something like this:

```sh
rq enqueue path.to.func --at-front --result-ttl 60 my_number=1 my_string=abc123
```

`at-front` and `result-ttl` are `enqueue` arguments while `my_number` and `my_string` will be passed to `func`.

Reply:
The problem with `literal_eval()` is that passing `abc123` fails, because it should be `'abc123'`. But passing this in e.g. bash or zsh fails as well because the shell parses `'abc123'` to `abc123`. So you would have to pass `\'abc123\'` or `'"abc123"'`. So I think the current version with `key:=json_value` or `:json_value` is more understandable and simpler to use.

> The goal is to be able to do something like this:
> 
> ```shell
> rq enqueue path.to.func --at-front --result-ttl 60 my_number=1 my_string=abc123
> ```
> 
> `at-front` and `result-ttl` are `enqueue` arguments while `my_number` and `my_string` will be passed to `func`.

After the latest commit that should work except that `my_number` will be parsed as a string so you have to change this to `my_number:=1`.

Reply:
Please write me if I should change `json.loads` to `ast.literal_eval`.

With that you would have the `:=` as well, but you would have access to for example sets or tuples.

Reply:
> Do you mind documenting this feature too so people know how to use it? The docs are under `/docs` directory.

No, but after we discussed how this should actually work.

Reply:
I think the best would be to support both `json.loads` and `ast.literal_eval` because `ast.literal_eval` is more functional but `json.loads` is more standardized. Maybe `ast.literal_eval` could be accessed with changeing `key#=value` and `#value`.